DylibBinder/InnerX.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using SwiftReflector.Inventory;
+
+namespace DylibBinder {
+	internal class InnerX {
+		// This class will be responsible for creating a dictionary
+		// telling us if a TypeDeclaration is on the top level and if not,
+		// which TypeDeclaration it is an innerType for
+
+		public InnerX ()
+		{
+		}
+
+		public Dictionary<string, List<ClassContents>> InnerXDict { get; } = new Dictionary<string, List<ClassContents>> ();
+
+		public Dictionary<string, List<ClassContents>> AddClassContentsList (params List<ClassContents>[] contents)
+		{
+			foreach (var classContentsList in contents) {
+				foreach (var classContent in classContentsList) {
+					AddItem (classContent);
+				}
+			}
+			return InnerXDict;
+		}
+
+		void AddItem (ClassContents c)
+		{
+			var nestingNames = c.Name.NestingNames.ToList ();
+			if (nestingNames.Count == 0)
+				return;
+
+			var nestingNameString = c.Name.ToFullyQualifiedName (true);
+
+			if (nestingNameString == null)
+				return;
+
+			if (nestingNames.Count == 1) {
+				AddKeyIfNotPresent (nestingNameString);
+				return;
+			}
+
+			var parentNestingNameString = GetParentNameString (nestingNameString);
+			if (parentNestingNameString == null)
+				return;
+
+			AddKeyIfNotPresent (parentNestingNameString);
+
+			var value = InnerXDict [parentNestingNameString];
+			value.Add (c);
+			InnerXDict [parentNestingNameString] = value;
+		}
+
+		void AddKeyIfNotPresent (string key)
+		{
+			if (InnerXDict.ContainsKey (key))
+				return;
+			InnerXDict.Add (key, new List<ClassContents> ());
+		}
```

Comment:
Lets use collections that will check for the presence rather than a list. https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-5.0

Reply:
Here I am using a dictionary \<string, List\<ClassContents\>\> and just checking if the dictionary has a specific key and creating that key if it does not so I don't think the hashset will apply here, correct?